File: api/python/yr/executor/function_handler.py
#!/usr/bin/env python3
# coding=UTF-8
# Copyright (c) Huawei Technologies Co., Ltd. 2024. All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""function handler"""

import logging
import os
import traceback
import inspect
from typing import List, Tuple
import uuid

from yr.libruntime_pb2 import InvokeType
from yr.npu_object import NpuObject
import yr
from yr.code_manager import CodeManager
from yr.common.utils import err_to_str
from yr.err_type import ErrorCode, ErrorInfo, ModuleCode
from yr.exception import YRInvokeError
from yr.executor.handler_intf import HandlerIntf
from yr.executor.instance_manager import InstanceManager
from yr.signature import recover_args
from yr.ds_tensor_client_manager import get_tensor_client

try:
    import torch
    import torch_npu
    from torch.npu import current_device

    TORCH_IMPORTED = True
except ImportError as import_err:
    TORCH_IMPORTED = False
    _import_error = import_err

USER_SHUTDOWN_FUNC_NAME = "__yr_shutdown__"
USER_BEFORE_SNAPSHOT_FUNC_NAME = "__yr_before_snapshot__"
USER_AFTER_SNAPSTART_FUNC_NAME = "__yr_after_snapstart__"

_logger = logging.getLogger(__name__)


def _store_tensor_to_ds(results, instance_function_name):
    def dev_mset_to_ds(tensor: torch.Tensor, instance_function_name):
        key = str(uuid.uuid4())
        ds_client = get_tensor_client()
        _logger.info(
            f"start dev mset, tensor key is {key}, function name: {instance_function_name}")
        failed_keys = ds_client.dev_mset([key], [tensor])
        if failed_keys:
            _logger.error(
                f"dev mset failed, failed keys is {failed_keys}, function name is {instance_function_name}")
            raise RuntimeError(
                f"dev_mset failed, failed keys: {failed_keys}, function name: {instance_function_name}")
        InstanceManager().store_tensor_in_local(tensor, key)
        return NpuObject(key, tensor.size(), tensor.dtype, tensor.device, os.getenv("YR_DS_ADDRESS"),
                         os.getenv("INSTANCE_ID"))

    if isinstance(results, torch.Tensor):
        return dev_mset_to_ds(results, instance_function_name)
    if isinstance(results, list):
        return [dev_mset_to_ds(result, instance_function_name) for result in results]
    if isinstance(results, tuple):
        return tuple(dev_mset_to_ds(result, instance_function_name) for result in results)
    if isinstance(results, set):
        return {dev_mset_to_ds(result, instance_function_name) for result in results}
    if isinstance(results, dict):
        return {k: dev_mset_to_ds(v, instance_function_name) for k, v in results.items()}
    return results


def _get_tensor_from_ds(*args, **kwargs):
    new_args = []
    new_kwargs = {}

    def dev_mget_from_ds(arg):
        if isinstance(arg, NpuObject):
            if not TORCH_IMPORTED:
                raise RuntimeError(f"import err: {_import_error}")
            ds_client = get_tensor_client()
            out_tensor = torch.zeros(size=arg.size, dtype=arg.dtype, device=f'npu:{current_device()}')
            _logger.info(f"start get tensor from ds, arg id is {arg.id}, current device is {current_device()}")
            failed_keys = ds_client.dev_mget([arg.id], [out_tensor])
            if failed_keys:
                _logger.info(f"dev_mget failed, arg id is {arg.id}, failed key is {failed_keys}")
                raise RuntimeError(f"dev_mget failed, failed keys: {failed_keys}")
            return out_tensor
        elif isinstance(arg, list):
            return [dev_mget_from_ds(item) for item in arg]
        elif isinstance(arg, tuple):
            return tuple(dev_mget_from_ds(item) for item in arg)
        elif isinstance(arg, set):
            return {dev_mget_from_ds(item) for item in arg}
        elif isinstance(arg, dict):
            return {key: dev_mget_from_ds(value) for key, value in arg.items()}
        return arg

    for arg in args:
        new_args.append(dev_mget_from_ds(arg))
    for k, v in kwargs.items():
        new_kwargs[k] = dev_mget_from_ds(v)
    return new_args, new_kwargs


class FunctionHandler(HandlerIntf):
    """
    Function handler class
    """

    def execute_function(
            self, func_meta, args: List, invoke_type, return_num: int,
            is_actor_async: bool) -> Tuple[List[object], ErrorInfo]:
        """execute function"""
        result = None
        try:
            if invoke_type == InvokeType.CreateInstance:
                self.__create_instance(func_meta, args)
            elif invoke_type == InvokeType.InvokeFunction:
                result = self.__invoke_instance(func_meta, args, is_actor_async)
            elif invoke_type == InvokeType.CreateInstanceStateless:
                self.__initialize_stateless_instance(func_meta)
            elif invoke_type == InvokeType.InvokeFunctionStateless:
                result = self.__invoke_stateless_function(func_meta, args)
            elif invoke_type == InvokeType.GetNamedInstanceMeta:
                result = InstanceManager().class_code
            elif invoke_type == InvokeType.DeleteRemoteTensor:
                result = self.__delete_tensors(args)
            else:
                msg = f"invalid invoke type {invoke_type}"
                _logger.warning(msg)
                return [], ErrorInfo(ErrorCode.ERR_EXTENSION_META_ERROR, ModuleCode.RUNTIME, msg)
        except Exception as err:
            _logger.warning("failed to execute user function, err: %s", err_to_str(err))
            if isinstance(err, YRInvokeError):
                result = [YRInvokeError(err.cause, traceback.format_exc())]
            else:
                result = [YRInvokeError(err, traceback.format_exc())]
            return result, ErrorInfo(ErrorCode.ERR_USER_FUNCTION_EXCEPTION, ModuleCode.RUNTIME,
                                     f"failed to execute user function, err: {repr(err)}")

        try:
            result_new = self.__check_return_list(result, return_num)
        except (TypeError, ValueError) as e:
            result_new = [YRInvokeError(e, traceback.format_exc())]
            _logger.error("Errors found during checking return values list, error: %s", e)
            return result_new, ErrorInfo(ErrorCode.ERR_USER_FUNCTION_EXCEPTION, ModuleCode.RUNTIME,
                                         f"failed to execute user function, err: {repr(e)}")

        return result_new, ErrorInfo()

    def shutdown(self, grace_period_second: int) -> ErrorInfo:
        """shutdown"""
        _logger.debug("Start to call user shutdown function __yr_shutdown__")
        instance = InstanceManager().instance()
        if instance is None:
            return ErrorInfo(
                ErrorCode.ERR_INNER_SYSTEM_ERROR,
                ModuleCode.RUNTIME,
                f"Failed to invoke instance function [{USER_SHUTDOWN_FUNC_NAME}], instance has not been initialized",
            )

        shutdown_func = None
        try:
            shutdown_func = getattr(instance, USER_SHUTDOWN_FUNC_NAME)
        except AttributeError:
            return ErrorInfo()
        try:
            shutdown_func(grace_period_second)
            _logger.info("Succeeded to call user shutdown function __yr_shutdown__")
        except Exception as e:
            _logger.exception(e)
            return ErrorInfo(ErrorCode.ERR_INNER_SYSTEM_ERROR, ModuleCode.RUNTIME, err_to_str(e))
        return ErrorInfo()

    def __before_snapshot(self, instance) -> ErrorInfo:
        """
        Call user-defined __yr_before_snapshot__ hook before taking snapshot.

        Args:
            instance: The user instance object.

        Returns:
            ErrorInfo: Error information if hook execution failed.
        """
        if instance is None:
            return ErrorInfo(
                ErrorCode.ERR_INNER_SYSTEM_ERROR,
                ModuleCode.RUNTIME,
                f"Failed to invoke instance function [{USER_BEFORE_SNAPSHOT_FUNC_NAME}], instance has not been initialized",
            )

        # Check if user defined the hook method
        if not hasattr(instance, USER_BEFORE_SNAPSHOT_FUNC_NAME):
            log.get_logger().debug(f"User hook {USER_BEFORE_SNAPSHOT_FUNC_NAME} not defined, skipping")
            return ErrorInfo()

        snapshot_hook = getattr(instance, USER_BEFORE_SNAPSHOT_FUNC_NAME)
        if not callable(snapshot_hook):
            log.get_logger().debug(f"User hook {USER_BEFORE_SNAPSHOT_FUNC_NAME} is not callable, skipping")
            return ErrorInfo()

        log.get_logger().debug(f"Start to call user snapshot hook {USER_BEFORE_SNAPSHOT_FUNC_NAME}")
        try:
            snapshot_hook()
            log.get_logger().info(f"Succeeded to call user snapshot hook {USER_BEFORE_SNAPSHOT_FUNC_NAME}")
        except Exception as e:
            log.get_logger().exception(e)
            return ErrorInfo(ErrorCode.ERR_INNER_SYSTEM_ERROR, ModuleCode.RUNTIME, err_to_str(e))
        return ErrorInfo()

    def __after_snapstarted(self, instance) -> ErrorInfo:
        """
        Call user-defined __yr_after_snapstart__ hook after restoring from snapshot.

        Args:
            instance: The user instance object.

        Returns:
            ErrorInfo: Error information if hook execution failed.
        """
        if instance is None:
            return ErrorInfo(
                ErrorCode.ERR_INNER_SYSTEM_ERROR,
                ModuleCode.RUNTIME,
                f"Failed to invoke instance function [{USER_AFTER_SNAPSTART_FUNC_NAME}], instance has not been initialized",
            )

        # Check if user defined the hook method
        if not hasattr(instance, USER_AFTER_SNAPSTART_FUNC_NAME):
            log.get_logger().debug(f"User hook {USER_AFTER_SNAPSTART_FUNC_NAME} not defined, skipping")
            return ErrorInfo()

        snapstart_hook = getattr(instance, USER_AFTER_SNAPSTART_FUNC_NAME)
        if not callable(snapstart_hook):
            log.get_logger().debug(f"User hook {USER_AFTER_SNAPSTART_FUNC_NAME} is not callable, skipping")
            return ErrorInfo()

        log.get_logger().debug(f"Start to call user snapstart hook {USER_AFTER_SNAPSTART_FUNC_NAME}")
        try:
            snapstart_hook()
            log.get_logger().info(f"Succeeded to call user snapstart hook {USER_AFTER_SNAPSTART_FUNC_NAME}")
        except Exception as e:
            log.get_logger().exception(e)
            return ErrorInfo(ErrorCode.ERR_INNER_SYSTEM_ERROR, ModuleCode.RUNTIME, err_to_str(e))
        return ErrorInfo()

    def before_snapshot(self) -> ErrorInfo:
        """
        Public method to trigger snapshot hook, called by libruntime before snapshot.

        Returns:
            ErrorInfo: Error information if hook execution failed.
        """
        log.get_logger().debug("Trigger before_snapshot hook")
        instance = InstanceManager().instance()
        return self.__before_snapshot(instance)

    def after_snapstart(self) -> ErrorInfo:
        """
        Public method to trigger snapstart hook, called by libruntime after restore.

        Returns:
            ErrorInfo: Error information if hook execution failed.
        """
        log.get_logger().debug("Trigger after_snapstart hook")
        instance = InstanceManager().instance()
        return self.__after_snapstarted(instance)

    def __create_instance(self, func_meta, args) -> None:
        _logger.info("%s" % func_meta)
        class_code = CodeManager().load_code(func_meta, True)
        if class_code is None:
            raise RuntimeError(f"Failed to load code from data system, code id: [{func_meta.codeID}]")
        InstanceManager().class_code = class_code
        InstanceManager().set_code_ref(func_meta.codeID, False)
        InstanceManager().is_async = func_meta.isAsync

        args, kwargs = self.__get_param(args)
        instance = class_code(*args, **kwargs)
        InstanceManager().init(instance)

    def __invoke_instance(self, func_meta, args, is_actor_async) -> object:
        enable_tensor_transport = func_meta.enableTensorTransport
        tensor_transport_target = func_meta.tensorTransportTarget
        args, kwargs = self.__get_param(args)

        if tensor_transport_target == 'npu':
            args, kwargs = _get_tensor_from_ds(*args, **kwargs)

        instance_function_name = func_meta.functionName
        instance = InstanceManager().instance()
        if instance is None:
            raise RuntimeError(
                f"Failed to invoke instance function [{instance_function_name}], instance has not been initialized")

        def sync_to_async(func):
            if inspect.iscoroutinefunction(func) or inspect.isasyncgenfunction(func):
                return func

            async def wrapper(*args, **kwargs):
                res = func(*args, **kwargs)
                if inspect.isawaitable(res):
                    return await res
                return res

            return wrapper

        func = getattr(instance, instance_function_name)
        if not func_meta.isAsync and is_actor_async:
            func = sync_to_async(func)
        results = func(*args, **kwargs)
        if enable_tensor_transport:
            return _store_tensor_to_ds(results, instance_function_name)
        return results

    def __delete_tensors(self, args):
        args, _ = self.__get_param(args)
        return InstanceManager().erase_tensors(args)

    def __initialize_stateless_instance(self, func_meta):
        if func_meta.initializerCodeID:
            initializer_code = CodeManager().load_code_from_datasystem(func_meta.initializerCodeID)
            initializer_code()

    def __invoke_stateless_function(self, func_meta, args):
        code = CodeManager().load_code(func_meta)
        if code is None:
            raise RuntimeError(f"Failed to load code from data system, code id: [{func_meta.codeID}]")

        args, kwargs = self.__get_param(args)
        result = code(*args, **kwargs)
        return result

    def __get_param(self, args: List) -> Tuple[object, object]:
        params = yr.serialization.Serialization().deserialize(args)
        args, kwargs = recover_args(params)
        return args, kwargs

    def __check_return_list(self, result, return_num: int):
        if isinstance(result, YRInvokeError) or return_num == 1:
            return [result] * return_num
        if return_num > 1:
            if not hasattr(result, "__len__"):
                raise TypeError(f"cannot unpack non-iterable {type(result)} object")
            if len(result) != return_num:
                raise ValueError(f"not enough values to unpack (expected {return_num}, got {len(result)})")
        return result

File: api/java/yr-runtime/src/main/java/com/yuanrong/executor/FunctionHandler.java
/*
 * Copyright (c) Huawei Technologies Co., Ltd. 2024-2024. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.yuanrong.executor;

import com.yuanrong.errorcode.ErrorCode;
import com.yuanrong.errorcode.ErrorInfo;
import com.yuanrong.errorcode.ModuleCode;
import com.yuanrong.errorcode.Pair;
import com.yuanrong.exception.YRException;
import com.yuanrong.exception.handler.filter.FilterFactory;
import com.yuanrong.exception.handler.traceback.StackTraceInfo;
import com.yuanrong.libruntime.generated.Libruntime;
import com.yuanrong.libruntime.generated.Libruntime.FunctionMeta;
import com.yuanrong.runtime.config.RuntimeContext;
import com.yuanrong.runtime.util.Utils;
import com.yuanrong.runtime.util.ExtClasspathLoader;
import com.yuanrong.serialization.Serializer;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.exc.MismatchedInputException;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.io.PrintWriter;
import java.io.StringWriter;
import java.lang.reflect.Constructor;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * FunctionHandler
 *
 * @since 2024/08/15
 */
public class FunctionHandler implements HandlerIntf {
    /**
     * String (clzName) => Object (instance)
     */
    public static final Map<String, Class<?>> classCache = new HashMap<String, Class<?>>();

    /**
     * String (clzName) => String (funcName) => Method (method id)
     */
    public static final Map<String, Map<String, Method>> methodCache = new HashMap<String, Map<String, Method>>();

    /**
     * CONSTRUCTOR_FUNC_NAME init value
     */
    public static final String CONSTRUCTOR_FUNC_NAME = "<init>";

    private static final Logger LOG = LoggerFactory.getLogger(FunctionHandler.class);

    private static final String USER_SHUTDOWN_FUNC_NAME = "yrShutdown";

    private static final String USER_SHUTDOWN_FUNC_SIGNATURE = "(I)V";

    private static final String USER_RECOVER_FUNC_NAME = "yrRecover";

    private static final String USER_RECOVER_FUNC_SIGNATURE = "()V";

    private static Object instance = null;

    private static String instanceClassName = "";

    private static Class<?> getClassThroughCache(String clzName) throws ClassNotFoundException {
        Class<?> res = classCache.get(clzName);
        if (res != null) {
            return res;
        }
        Class<?> clz = ExtClasspathLoader.getFunctionClassLoader().loadClass(clzName);
        classCache.put(clzName, clz);
        return clz;
    }

    /**
     * Execute function
     *
     * @param meta the meta
     * @param type the type
     * @param args the args
     * @return the return value, in ByteBuffer type, may need to release bytebuffer
     * `buffer.clear()`
     * @throws Exception the exception
     */
    @Override
    public ReturnType execute(FunctionMeta meta, Libruntime.InvokeType type, List<ByteBuffer> args) throws Exception {
        LOG.info("executing udf methods, current type: {}", type);

        switch (type) {
            case CreateInstance:
                // $FALL-THROUGH$
            case CreateInstanceStateless:
                LOG.debug("Invoking udf method matched, case create");
                return create(meta, args);
            case InvokeFunction:
                // $FALL-THROUGH$
            case InvokeFunctionStateless:
                LOG.debug("Invoking udf method matched, case invoke");
                return invoke(meta, args);
            default:
                LOG.debug("Invoking udf method matched, case dft");
                return new ReturnType(ErrorCode.ERR_INCORRECT_INVOKE_USAGE, "invalid invoke type");
        }
    }

    /**
     * Shutdown the instance gracefully.
     *
     * @param gracePeriodSeconds the time to wait for the instance to shutdown gracefully.
     * @return ErrorInfo, the ErrorInfo of the execution of shutdown function.
     */
    @Override
    public ErrorInfo shutdown(int gracePeriodSeconds) {
        if (instance == null) {
            return new ErrorInfo(ErrorCode.ERR_INNER_SYSTEM_ERROR, ModuleCode.RUNTIME,
                "Failed to invoke instance function [" + USER_SHUTDOWN_FUNC_NAME
                    + "], instance has not been initialized");
        }
        Class<?> clz = null;
        Method method = null;
        try {
            LOG.debug("Start to get shutdown method {} from class {}", USER_SHUTDOWN_FUNC_NAME, instanceClassName);
            clz = getClassThroughCache(instanceClassName);
            Class<?>[] paramTypes = Utils.getParameterTypeFromSignature(clz, USER_SHUTDOWN_FUNC_NAME,
                USER_SHUTDOWN_FUNC_SIGNATURE);
            method = clz.getMethod(USER_SHUTDOWN_FUNC_NAME, paramTypes);
        } catch (ClassNotFoundException | NoSuchMethodException | IllegalArgumentException e) {
            LOG.debug("Cannot get {} from the class cache, exception: {}", USER_SHUTDOWN_FUNC_NAME,
                getCausedByString(e));
            return new ErrorInfo();
        }
        LOG.debug("Succeeded to get shutdown method {} from class {}", USER_SHUTDOWN_FUNC_NAME, instanceClassName);

        try {
            method.invoke(instance, gracePeriodSeconds);
        } catch (IllegalAccessException | IllegalArgumentException | InvocationTargetException e) {
            LOG.debug("Failed to invoke user {} function, exception: {}", USER_SHUTDOWN_FUNC_NAME,
                getCausedByString(e));
            return new ErrorInfo(ErrorCode.ERR_INNER_SYSTEM_ERROR, ModuleCode.RUNTIME,
                "Failed to invoke user " + USER_SHUTDOWN_FUNC_NAME + " function");
        }

        LOG.debug("Succeeded to invoke user {} function", USER_SHUTDOWN_FUNC_NAME);
        return new ErrorInfo(ErrorCode.ERR_OK, ModuleCode.RUNTIME, "");
    }

    /**
     * Serializes the instance of the CodeExecutor class and returns a Pair
     * containing the serialized byte arrays of the instance and the class name.
     *
     * @param instanceID the ID of the instance to be dumped
     * @return a Pair containing the serialized byte arrays of the instance and the
     * class name
     * @throws JsonProcessingException if there is an error during serialization
     */
    @Override
    public Pair<byte[], byte[]> dumpInstance(String instanceID) throws JsonProcessingException {
        byte[] instanceBytes = Serializer.serialize(FunctionHandler.instance);
        byte[] clzNameBytes = Serializer.serialize(FunctionHandler.instanceClassName);
        LOG.debug("Succeeded to DumpInstance(clzName: {}), instanceID: {}", FunctionHandler.instanceClassName,
            instanceID);
        return new Pair<>(instanceBytes, clzNameBytes);
    }

    /**
     * Loads an instance of a class from serialized byte arrays.
     *
     * @param instanceBytes the serialized byte array representing the instance
     * @param clzNameBytes the serialized byte array representing the class name
     * @throws IOException if there is an error reading the byte arrays
     * @throws ClassNotFoundException if the class specified by the class name is not found
     */
    @Override
    public void loadInstance(byte[] instanceBytes, byte[] clzNameBytes) throws IOException, ClassNotFoundException {
        String clzName = (String) Serializer.deserialize(clzNameBytes, String.class);
        Class<?> clz = getClassThroughCache(clzName);
        Object inst;
        try {
            inst = Serializer.deserialize(instanceBytes, clz);
        } catch (MismatchedInputException e) {
            throw new IOException("Catch MismatchedInputException while running LoadInstance, please check "
                + "whether the getter/setter methods in user class comply with the Java Beans naming rules. "
                + "Exception:", e);
        }
        LOG.debug("Succeeded to LoadInstance, clzName: {}", clzName);
        instance = inst;
        instanceClassName = clzName;
    }

    /**
     * Recover the instance.
     *
     * @return ErrorInfo, the ErrorInfo of the execution of recover function.
     */
    @Override
    public ErrorInfo recover() {
        if (instance == null) {
            return new ErrorInfo(ErrorCode.ERR_INNER_SYSTEM_ERROR, ModuleCode.RUNTIME,
                "Failed to invoke instance function [" + USER_RECOVER_FUNC_NAME
                    + "], instance has not been initialized");
        }
        Class<?> clz = null;
        Method method = null;
        try {
            LOG.debug("Start to get recover method {} from class {}", USER_RECOVER_FUNC_NAME, instanceClassName);
            clz = getClassThroughCache(instanceClassName);
            Class<?>[] paramTypes = Utils.getParameterTypeFromSignature(clz, USER_RECOVER_FUNC_NAME,
                USER_RECOVER_FUNC_SIGNATURE);
            method = clz.getMethod(USER_RECOVER_FUNC_NAME, paramTypes);
        } catch (ClassNotFoundException | NoSuchMethodException | IllegalArgumentException e) {
            LOG.debug("Cannot get {} from the class cache, exception: {}", USER_RECOVER_FUNC_NAME,
                getCausedByString(e));
            return new ErrorInfo(ErrorCode.ERR_OK, ModuleCode.RUNTIME, "");
        }
        LOG.debug("Succeeded to get recover method {} from class {}", USER_RECOVER_FUNC_NAME, instanceClassName);

        try {
            method.invoke(instance);
        } catch (IllegalAccessException | IllegalArgumentException | InvocationTargetException e) {
            LOG.debug("Failed to invoke user {} function, exception: {}", USER_RECOVER_FUNC_NAME,
                getCausedByString(e));
            return new ErrorInfo(ErrorCode.ERR_INNER_SYSTEM_ERROR, ModuleCode.RUNTIME,
                "Failed to invoke user " + USER_RECOVER_FUNC_NAME + " function");
        }

        LOG.debug("Succeeded to invoke user {} function", USER_RECOVER_FUNC_NAME);
        return new ErrorInfo(ErrorCode.ERR_OK, ModuleCode.RUNTIME, "");
    }

    /**
     * Create ReturnType
     *
     * @param meta the meta
     * @param args the args
     * @return ReturnType
     */
    private static ReturnType create(FunctionMeta meta, List<ByteBuffer> args) {
        String clzName = meta.getClassName();
        String funcName = meta.getFunctionName();
        String signature = meta.getSignature();
        LOG.info("Creating for udf methods, clz({}), func({}), sig({})", clzName, funcName, signature);
        try {
            constructObject(clzName, funcName, signature, args);
        } catch (ClassNotFoundException | InstantiationException | IllegalAccessException | IllegalArgumentException
                 | InvocationTargetException | IOException e) {
            LOG.error("failed to execute udf methods, clz({}), func({}), sig({}), reason: {}", clzName, funcName,
                signature, e);
            return new ReturnType(ErrorCode.ERR_USER_FUNCTION_EXCEPTION, ModuleCode.RUNTIME_CREATE,
                "failed to create instance due to the cause: \n" + getCausedByString(e));
        }
        LOG.info("successfully finish the create job with clz({}), func({}), sig({})", clzName, funcName, signature);
        return new ReturnType(ErrorCode.ERR_OK, "");
    }

    /**
     * construct an object, and cache it in the `instance`
     *
     * @param clzName the class name, required
     * @param funcName the function name, if is <init>, will use constructors
     * @param signature the constructor signature, optional, will use default
     * constructor if missing
     * @param args the constructor args, will be passed to the constructor
     * @return the new object
     * @throws ClassNotFoundException the ClassNotFoundException
     * @throws InstantiationException the InstantiationException
     * @throws IllegalAccessException the IllegalAccessException
     * @throws IllegalArgumentException the IllegalArgumentException
     * @throws InvocationTargetException the InvocationTargetException
     * @throws IOException the IOException
     */
    private static Object constructObject(String clzName, String funcName, String signature, List<ByteBuffer> args)
        throws ClassNotFoundException, InstantiationException, IllegalAccessException, IllegalArgumentException,
               InvocationTargetException, IOException {
        Class<?> clz = ExtClasspathLoader.getFunctionClassLoader().loadClass(clzName);
        LOG.debug("constructing clz({}), func({}), sig({}), get class ok: {}", clzName, funcName, signature, clz);
        // Match constructor
        Constructor<?> constructor = null;
        for (Constructor<?> cs : clz.getConstructors()) {
            // if funcName is not a constructor, will just use the default one
            // if signature missed, use the first one
            LOG.debug("Creating for udf methods, current cs: {}", cs);
            if (signature == null || signature.isEmpty() || !CONSTRUCTOR_FUNC_NAME.equals(funcName) || signature.equals(
                Utils.getMethodSignature(cs))) {
                constructor = cs;
                break;
            }
        }
        // failed to find anyone match the requirements
        if (constructor == null) {
            LOG.error("Failed to get constructor");
            throw new RuntimeException("failed to find constructor with specified signature: (" + signature + ")");
        }

        LOG.debug("successfully get constructor: {}", constructor);
        Object[] objArgs = Utils.deserializeObjects(Arrays.asList(constructor.getParameterTypes()), args).toArray();
        LOG.debug("successfully deserialize objects: {}", objArgs.length);

        // deserialize the params to objects, and new instance
        Object inst = constructor.newInstance(objArgs);
        LOG.debug("successfully construct obj: {}", inst);
        FunctionHandler.instance = inst;
        FunctionHandler.instanceClassName = clzName;
        return inst;
    }

    private static ReturnType invoke(FunctionMeta meta, List<ByteBuffer> args) {
        LOG.debug("Invoking udf methods, meta: {}", meta);
        String clzName = meta.getClassName();
        String funcName = meta.getFunctionName();
        String signature = meta.getSignature();
        LOG.debug("Invoking udf methods, clz({}), func({}), sig({})", clzName, funcName, signature);
        Class<?>[] paramTypes = null;
        Method method = null;

        try {
            // New such class in cache, just new hash map
            if (!methodCache.containsKey(clzName)) {
                methodCache.put(clzName, new HashMap<String, Method>());
            }
            // No such function in cache, new function
            if (!methodCache.get(clzName).containsKey(funcName)) {
                Class<?> clz = getClassThroughCache(clzName);
                paramTypes = Utils.getParameterTypeFromSignature(clz, funcName, signature);
                method = clz.getMethod(funcName, paramTypes);
                methodCache.get(clzName).put(funcName, method);
            }
            method = method == null ? methodCache.get(clzName).get(funcName) : method;
            paramTypes = paramTypes == null ? Utils.getParameterTypeFromSignature(getClassThroughCache(clzName),
                funcName, signature) : paramTypes;
            if (instance == null) {
                throw new RuntimeException(
                    "failed to find instance object when invoke method (" + meta.toString() + ")");
            }
            List<Object> argObjs = Utils.deserializeObjects(Arrays.asList(paramTypes), args);
            Object retValue = method.invoke(instance, argObjs.toArray());
            LOG.debug("Succeeded to invoke method, class({}), function name({}).", clzName, funcName);
            return new ReturnType(ErrorCode.ERR_OK, Serializer.serialize(retValue));
        } catch (ClassNotFoundException | IllegalAccessException | IllegalArgumentException
                 | InvocationTargetException | IOException | NoSuchMethodException e) {
            return processException(e, clzName, funcName);
        }
    }

    private static ReturnType processException(Exception e, String clzName, String funcName) {
        LOG.debug("Get exception when invoke for {}.{}, exception type:{}", clzName, funcName, e.getClass().getName());
        // if throwable is not InvocationTargetException, consider it as inner system exception
        if (!(e instanceof InvocationTargetException)) {
            LOG.error("Get not InvocationTargetException exception when invoke for {}.{}, exception:{}, message:{}",
                clzName, funcName, e.toString(), getCausedByString(e));
            return new ReturnType(ErrorCode.ERR_INNER_SYSTEM_ERROR, ModuleCode.RUNTIME_INVOKE, getCausedByString(e));
        }
        InvocationTargetException targetException = (InvocationTargetException) e;
        Throwable throwable = targetException.getTargetException();
        LOG.debug("Get throwable after peel off InvocationTargetException when invoke for {}.{}, {}", clzName, funcName,
            throwable.toString());

        // if throwable is not YRException, consider it as user code origin exception
        if (!(throwable instanceof YRException)) {
            FilterFactory filterFactory = new FilterFactory(throwable, true);
            StackTraceInfo stackTraceInfo = filterFactory.generateStackTraceInfo(clzName, funcName);
            LOG.error("Get origin exception when invoke for {}.{}, throwable:{}, stackTraceInfo:{}", clzName, funcName,
                throwable, stackTraceInfo.toString());
            List<StackTraceInfo> stackTraces = new ArrayList<StackTraceInfo>();
            stackTraces.add(stackTraceInfo);
            String errorMsg = "exception occurred when processing user code, runtimeId: " + RuntimeContext.runtimeId
                + ", throwable: " + throwable;
            return new ReturnType(ErrorCode.ERR_USER_FUNCTION_EXCEPTION, ModuleCode.RUNTIME_INVOKE, errorMsg,
                stackTraces);
        }

        // if throwable is YRException, and it is also an user function exception, then process it
        LOG.error("Get throwable from YRException when invoke for {}.{}, throwable:{}", clzName, funcName,
            throwable.toString());
        if (ErrorCode.ERR_USER_FUNCTION_EXCEPTION.equals(((YRException) throwable).getErrorCode())) {
            LOG.error("Get YRException exception when invoke for {}.{}, exception class:{}", clzName, funcName,
                e.getClass().getName());
            YRException yrException = (YRException) throwable;
            // if Exception is YRException and exception is caused by user code, then get exceptions in it and
            // join stacktrace of YRException and inner exception stacktraces
            LOG.debug("Get exceptions stackTraces:{} in YRException when invoke for {}.{}",
                Arrays.asList(yrException.getStackTrace()).toArray(), clzName, funcName);
            FilterFactory filterFactory = new FilterFactory(yrException.getCause(), false);
            StackTraceInfo stackTraceInfo = filterFactory.generateStackTraceInfo(clzName, funcName);
            LOG.debug("Get {}.{} result, YRException stackTraces:{}, innerExp.stackTraceInfo:{}, ", clzName,
                funcName, Arrays.asList(yrException.getStackTrace()).toArray(), stackTraceInfo.toString());
            List<StackTraceElement> finalElements = appendStackTraceArrayToList(stackTraceInfo.getStackTraceElements(),
                yrException.getStackTrace());

            LOG.debug("print final element:{} for {}.{}", finalElements, clzName, funcName);
            stackTraceInfo.setStackTraceElements(finalElements);
            List<StackTraceInfo> stackTraces = new ArrayList<StackTraceInfo>();
            stackTraces.add(stackTraceInfo);
            String errorMsg = "exception occurred when processing user code, runtimeId: "
                    + RuntimeContext.runtimeId + ", throwable: " + throwable;
            return new ReturnType(ErrorCode.ERR_USER_FUNCTION_EXCEPTION, ModuleCode.RUNTIME_INVOKE, errorMsg,
                stackTraces);
        }

        return new ReturnType(ErrorCode.ERR_USER_FUNCTION_EXCEPTION, ModuleCode.RUNTIME_INVOKE,
            getCausedByString(throwable));
    }

    private static List<StackTraceElement> appendStackTraceArrayToList(List<StackTraceElement> list,
        StackTraceElement[] stackTraceArray) {
        if (list.isEmpty()) {
            return new ArrayList<>(Arrays.asList(stackTraceArray));
        }
        if (stackTraceArray.length > 0) {
            List<StackTraceElement> stackTraceElementList = new ArrayList<>(Arrays.asList(stackTraceArray));
            list.addAll(stackTraceElementList);
            return list;
        }
        return new ArrayList<>();
    }

    private static String getCausedByString(Throwable throwable) {
        StringWriter stringWriter = new StringWriter();
        PrintWriter printWriter = new PrintWriter(stringWriter);
        throwable.printStackTrace(printWriter);
        return stringWriter.toString();
    }
}